File: frontend/src/app/services/cuenta.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Cuenta {
  id: number;
  nombre: string;
  tipo: string;
  saldo: number;
  tasaAnual: number;
  tipoTasa: string;
  diaDeCorteMensual: number | null;
  notas: string;
  rendimientoAnual: number;
  rendimientoMensual: number;
  rendimientoDiario: number;
  fechaCreacion: string;
}

export interface CrearCuenta {
  nombre: string;
  tipo: string;
  saldo: number;
  tasaAnual: number;
  tipoTasa: string;
  diaDeCorteMensual: number | null;
  notas: string;
}

@Injectable({ providedIn: 'root' })
export class CuentaService {
  private apiUrl = `${environment.apiUrl}/cuentas`;

  constructor(private http: HttpClient) {}

  obtenerTodas(): Observable<Cuenta[]> {
    return this.http.get<Cuenta[]>(this.apiUrl);
  }

  crear(dto: CrearCuenta): Observable<Cuenta> {
    return this.http.post<Cuenta>(this.apiUrl, dto);
  }

  actualizarSaldo(id: number, nuevoSaldo: number): Observable<Cuenta> {
    return this.http.put<Cuenta>(`${this.apiUrl}/${id}/saldo`, { nuevoSaldo });
  }

  eliminar(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }
}

File: PesoClaro.API/Services/CuentaService.cs
using Microsoft.EntityFrameworkCore;
using PesoClaro.API.Data;
using PesoClaro.API.DTOs;
using PesoClaro.API.Models;

namespace PesoClaro.API.Services;

public class CuentaService
{
    private readonly PesoClaroContext _context;

    public CuentaService(PesoClaroContext context)
    {
        _context = context;
    }

    public async Task<CuentaResponseDTO> CrearAsync(int usuarioId, CrearCuentaDTO dto)
    {
        var cuenta = new Cuenta
        {
            Nombre = dto.Nombre,
            Tipo = dto.Tipo,
            Saldo = dto.Saldo,
            TasaAnual = dto.TasaAnual,
            TipoTasa = dto.TipoTasa,
            DiaDeCorteMensual = dto.DiaDeCorteMensual,
            Notas = dto.Notas,
            UsuarioId = usuarioId
        };

        _context.Cuentas.Add(cuenta);
        await _context.SaveChangesAsync();

        return MapearACuentaResponse(cuenta);
    }

    public async Task<List<CuentaResponseDTO>> ObtenerTodasAsync(int usuarioId)
    {
        var cuentas = await _context.Cuentas
            .Where(c => c.UsuarioId == usuarioId && c.Activa)
            .OrderBy(c => c.Tipo)
            .ToListAsync();

        return cuentas.Select(MapearACuentaResponse).ToList();
    }

    public async Task<CuentaResponseDTO?> ActualizarSaldoAsync(
        int cuentaId, int usuarioId, ActualizarSaldoDTO dto)
    {
        var cuenta = await _context.Cuentas
            .FirstOrDefaultAsync(c => c.Id == cuentaId && c.UsuarioId == usuarioId && c.Activa);

        if (cuenta == null) return null;

        cuenta.Saldo = dto.NuevoSaldo;
        await _context.SaveChangesAsync();

        return MapearACuentaResponse(cuenta);
    }

    public async Task<bool> EliminarAsync(int cuentaId, int usuarioId)
    {
        var cuenta = await _context.Cuentas
            .FirstOrDefaultAsync(c => c.Id == cuentaId && c.UsuarioId == usuarioId && c.Activa);

        if (cuenta == null) return false;

        cuenta.Activa = false;
        await _context.SaveChangesAsync();

        return true;
    }

    private CuentaResponseDTO MapearACuentaResponse(Cuenta cuenta)
    {
        var rendimientoAnual = cuenta.TipoTasa == "Rendimiento"
            ? cuenta.Saldo * cuenta.TasaAnual / 100
            : -(cuenta.Saldo * cuenta.TasaAnual / 100);

        return new CuentaResponseDTO
        {
            Id = cuenta.Id,
            Nombre = cuenta.Nombre,
            Tipo = cuenta.Tipo,
            Saldo = cuenta.Saldo,
            TasaAnual = cuenta.TasaAnual,
            TipoTasa = cuenta.TipoTasa,
            DiaDeCorteMensual = cuenta.DiaDeCorteMensual,
            Notas = cuenta.Notas,
            RendimientoAnual = Math.Round(rendimientoAnual, 2),
            RendimientoMensual = Math.Round(rendimientoAnual / 12, 2),
            RendimientoDiario = Math.Round(rendimientoAnual / 365, 2),
            FechaCreacion = cuenta.FechaCreacion
        };
    }
}